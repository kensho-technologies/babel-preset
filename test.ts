import * as fs from 'fs'

import {test, expect} from '@jest/globals'
import {PluginItem, transform, TransformOptions} from '@babel/core'

import preset from '.'

// the 'test' environment is excluded because it targets the currently running version of Node
// since our CI runs multiple Node versions, a snapshot for 'test' might not match one of them
const NON_TEST_ENVIRONMENTS = [
  'development',
  'production',
  'esm',
  'cjs',
  'development-modern',
  'production-modern',
]

expect.addSnapshotSerializer({
  test: (val: unknown) => typeof val === 'string',
  print: (val: string) => val,
})

function macro(title: string, fixture: string, presetOptions = {}): void {
  test(`${title} given ${JSON.stringify(presetOptions)}`, () => {
    const file = `${__dirname}/fixtures/${fixture}`
    const input = fs.readFileSync(file, 'utf8')
    expect(input).toMatchSnapshot('input')

    const resolvedPresetOptions = {
      ...presetOptions,
      configPath: `${__dirname}/fixtures`,
    }

    NON_TEST_ENVIRONMENTS.forEach((envName) => {
      const presets: PluginItem[] = [[preset, resolvedPresetOptions]]
      const options: TransformOptions = {
        envName,
        presets,
        filename: `/${fixture}`,
        babelrc: false,
      }
      const result = transform(input, options)
      expect(result?.code).toMatchSnapshot(`output (${envName})`)
    })
  })
}

macro('transpiles ES2020+ syntax', 'syntax.js')
macro('transpiles ES2020+ syntax', 'syntax.js', {targets: 'Chrome 88'})
macro('transpiles ES2020+ syntax', 'syntax.js', {loose: false})
macro('transpiles ES2020+ syntax', 'syntax.js', {runtime: false})
macro('transpiles ES2020+ syntax', 'syntax.js', {modules: 'commonjs'})

macro('transpiles React', 'react.js')
macro('transpiles React with classic runtime', 'react-classic.js', {react: {runtime: 'classic'}})
macro('transpiles TSX', 'typescript.tsx')
macro('transpiles Emotion', 'emotion.js', {emotion: true})
macro('transpiles Emotion with classic runtime', 'emotion-classic.js', {
  emotion: true,
  react: {runtime: 'classic'},
})

macro('replaces generic polyfill with env-targeted polyfills', 'polyfills.js')

macro('transpiles ESM in node_modules', 'vendor/node_modules/my-pkg/index.js')
macro('transpiles CJS in node_modules', 'vendor/node_modules/my-pkg/cjs.js')
macro('avoids transpiling known precompiled packages', 'vendor/node_modules/react/index.js')
