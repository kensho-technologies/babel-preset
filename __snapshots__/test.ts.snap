// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transpiles ES2020+ syntax given {"targets":"last 2 Chrome versions"}: input 1`] = `
import defaultImport, {namedImport} from 'foo'
import * as namespaceImport from 'bar'

const dynamicImport = import('baz')

class Foo {
  static bar = 'abc'
  baz = (x, y) => x({...y})
}

const obj = {a: 1, b: 2}

const spread = {...obj, b: 2, c: 3}
const {a, ...rest} = spread

const nullishCoalescing = obj ?? 1
const optionalChaining = obj?.b

export default obj
export {obj}

`;

exports[`transpiles ES2020+ syntax given {"targets":"last 2 Chrome versions"}: output (production) 1`] = `
import defaultImport, { namedImport } from 'foo';
import * as namespaceImport from 'bar';
const dynamicImport = import('baz');

class Foo {
  constructor() {
    this.baz = (x, y) => x({ ...y
    });
  }

}

Foo.bar = 'abc';
const obj = {
  a: 1,
  b: 2
};
const spread = { ...obj,
  b: 2,
  c: 3
};
const {
  a,
  ...rest
} = spread;
const nullishCoalescing = obj != null ? obj : 1;
const optionalChaining = obj == null ? void 0 : obj.b;
export default obj;
export { obj };
`;

exports[`transpiles ES2020+ syntax given {"targets":"last 2 Chrome versions"}: output (production-modern) 1`] = `
import defaultImport, { namedImport } from 'foo';
import * as namespaceImport from 'bar';
const dynamicImport = import('baz');

class Foo {
  constructor() {
    this.baz = (x, y) => x({ ...y
    });
  }

}

Foo.bar = 'abc';
const obj = {
  a: 1,
  b: 2
};
const spread = { ...obj,
  b: 2,
  c: 3
};
const {
  a,
  ...rest
} = spread;
const nullishCoalescing = obj != null ? obj : 1;
const optionalChaining = obj == null ? void 0 : obj.b;
export default obj;
export { obj };
`;

exports[`transpiles ES2020+ syntax given {}: input 1`] = `
import defaultImport, {namedImport} from 'foo'
import * as namespaceImport from 'bar'

const dynamicImport = import('baz')

class Foo {
  static bar = 'abc'
  baz = (x, y) => x({...y})
}

const obj = {a: 1, b: 2}

const spread = {...obj, b: 2, c: 3}
const {a, ...rest} = spread

const nullishCoalescing = obj ?? 1
const optionalChaining = obj?.b

export default obj
export {obj}

`;

exports[`transpiles ES2020+ syntax given {}: output (production) 1`] = `
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import defaultImport, { namedImport } from 'foo';
import * as namespaceImport from 'bar';
var dynamicImport = import('baz');

var Foo = function Foo() {
  this.baz = function (x, y) {
    return x(Object.assign({}, y));
  };
};

Foo.bar = 'abc';
var obj = {
  a: 1,
  b: 2
};
var spread = Object.assign({}, obj, {
  b: 2,
  c: 3
});

var a = spread.a,
    rest = _objectWithoutPropertiesLoose(spread, ["a"]);

var nullishCoalescing = obj != null ? obj : 1;
var optionalChaining = obj == null ? void 0 : obj.b;
export default obj;
export { obj };
`;

exports[`transpiles ES2020+ syntax given {}: output (production-modern) 1`] = `
import defaultImport, { namedImport } from 'foo';
import * as namespaceImport from 'bar';
const dynamicImport = import('baz');

class Foo {
  constructor() {
    this.baz = (x, y) => x({ ...y
    });
  }

}

Foo.bar = 'abc';
const obj = {
  a: 1,
  b: 2
};
const spread = { ...obj,
  b: 2,
  c: 3
};
const {
  a,
  ...rest
} = spread;
const nullishCoalescing = obj != null ? obj : 1;
const optionalChaining = obj == null ? void 0 : obj.b;
export default obj;
export { obj };
`;
